OpenGL / 3D view refactor

neofelis2X opened the pull request:
Hi,

it is long overdue that I push this work on the 3D view. It reorganises / moves / rewrites most of the code that handles the opengl canvas. It does **not** enable _pyglet 2_, everything is still based on _pyglet 1.5_. But it makes all the foundational changes necessary to use modern OpenGL (3.3+).

That incldues:
- All opengl calls are now inside the gl package
    - they were all over the place and very hard to work with before
- Move duplicated code to canvas class and reuse it
    - stlview and gcodeview are not that different after all
- New camera class based on vectors
    - no more dolly-zoom by scaling the whole scene...
- Keyboard shortcuts organised in keyboard input class
- Fix various bugs on macOS / highDPI displays / mouse cursor
- Implement optional frametime counter 
- Define all 3D objects as actors (gl/actors)
- Draw the gridlines starting from center
- Add indicator to grid which indicates the origin point (0, 0)
- Gcode plater and Stl plater inherit _background color_ from settings
- Grid and focus outline change color based on _background_
- Add 3D testfiles in different formats
- Prepare some to code for vertexbuffers and shaders

I know this are a lot of changes and all in one PR. It has taken me a while to untangle it all and I have changed bit by bit, going through it all. In my opinion this was necessary to get a more modern starting point to work with. Splitting it into smaller PRs would be a hindrance. On the other hand, the following PR for pyglet 2 should be more concise.

Note 1: When the diff says that there are 50.000 new lines of code then that is because I added a .gcode testfile.
Note 2: I already have a branch where everything runs on _pyglet 2_ and it's working really well. But still work in progress.

Fixes #746
Fixes #1035
Fixes #1301

I would be happy if you can test it on your respective operating system. On macOS it runs really well and I also tested it on windows and it looked good. Any input is welcome!

DivingDuck commented:
Wow, this is an impressive work with a lot of code reorganization. 
I took only a quick look on it and feel like the GL part this is now better readable than before as I always was in trouble finding the right code on all the different locations. Not saying that I understand a lot of the GL part at all. Great work.

I did some quick test and this are my observations so fare:

- Main window, Gcode view: 
   - Mouse scroll wheel for zooming is not working
   - Fit model in window position is not correct (this behavior exists already before)

- Plater and Gcode-plater tool  (-->Tools -->Plater and -->Gcode-Plater)
   - you solved the 100% cpu load problem when running the plater tool for stl files
   - plater and gcode-plater
      - Mouse scroll wheel for zooming is not working
      - keyboard shortcut F for fit to window do only reset the view
   - Plater
      - got an error when I try to load a stl file:
```
Plate function activated
Not an ascii stl solid - attempting to parse as binary
OpenGL failed, disabling it:
Traceback (most recent call last):
  File "printrun\gl\panel.py", line 206, in processPaintEvent
  File "printrun\gl\panel.py", line 370, in DrawCanvas
  File "printrun\stlview.py", line 142, in draw_objects
  File "printrun\gl\panel.py", line 389, in transform_and_draw
  File "pyglet\graphics\__init__.py", line 564, in draw
  File "pyglet\graphics\__init__.py", line 493, in <lambda>
  File "pyglet\graphics\vertexdomain.py", line 292, in draw
  File "pyglet\graphics\vertexbuffer.py", line 398, in bind
  File "pyglet\gl\lib_wgl.py", line 103, in __call__
  File "pyglet\gl\lib.py", line 107, in errcheck
pyglet.gl.lib.GLException: b'ung\xfcltiger Wert'
```

- I saw the color of the grid have changed from green to something grayish and recognize that I can't change the color of the grid anymore. It looks like we lost the color setup for this since quite some time (what I thought we had before if I remember correct). This is for all model views.
- Looking in the code I recognize that there are missing some translation strings for logging that I had updated in the past. I will look on that later once more in the code review.

Again, you did a great job and I really like your improvements. 👍

rockstorm101 commented:
Thanks a lot for this @neofelis2X. I'll try and do some tests during the next two weeks.

> Note 2: I already have a branch where everything runs on _pyglet 2_ and it's working really well. But still work in progress.

Can't wait for those changes to be merged!

First test I tried. Playing around with the Plater (Tools -> Plater) failed with the following error. Plater window would not even show.
```
$ ./pronterface.py 
WARNING:root:Memory-efficient GCoder implementation unavailable: cannot import name 'gcoder_line' from 'printrun' (/home/rock/Projects/printrun/klimentrun/printrun/__init__.py)
3D view mode requested, but we failed to initialize it.
Falling back to 2D view, and here is the backtrace:
Traceback (most recent call last):
  File "/home/rock/Projects/printrun/klimentrun/printrun/gui/viz.py", line 69, in __init__
    import printrun.gcview
  File "/home/rock/Projects/printrun/klimentrun/printrun/gcview.py", line 23, in <module>
    from .gl.panel import wxGLPanel
  File "/home/rock/Projects/printrun/klimentrun/printrun/gl/panel.py", line 36, in <module>
    from pyglet.gl import GL_LIGHTING, GL_LIGHT0, GL_LIGHT1, GL_POSITION, \
    ...<4 lines>...
        glPushMatrix, glPopMatrix, glMultMatrixd, glMatrixMode
ImportError: cannot import name 'GL_LIGHTING' from 'pyglet.gl' (/usr/lib/python3/dist-packages/pyglet/gl/__init__.py)

Could not load 3D viewer for plater:
Traceback (most recent call last):
  File "/home/rock/Projects/printrun/klimentrun/printrun/stlplater.py", line 42, in <module>
    from printrun import stlview
  File "/home/rock/Projects/printrun/klimentrun/printrun/stlview.py", line 26, in <module>
    from .gl.panel import wxGLPanel
  File "/home/rock/Projects/printrun/klimentrun/printrun/gl/panel.py", line 36, in <module>
    from pyglet.gl import GL_LIGHTING, GL_LIGHT0, GL_LIGHT1, GL_POSITION, \
    ...<4 lines>...
        glPushMatrix, glPopMatrix, glMultMatrixd, glMatrixMode
ImportError: cannot import name 'GL_LIGHTING' from 'pyglet.gl' (/usr/lib/python3/dist-packages/pyglet/gl/__init__.py)

Plate function activated
Traceback (most recent call last):
  File "/home/rock/Projects/printrun/klimentrun/printrun/pronterface.py", line 587, in plate
    plater.StlPlater(size = (800, 580), callback = self.platecb,
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                     parent = self,
                     ^^^^^^^^^^^^^^
    ...<2 lines>...
                     simarrange_path = self.settings.simarrange_path,
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                     antialias_samples = int(self.settings.antialias3dsamples)).Show()
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rock/Projects/printrun/klimentrun/printrun/objectplater.py", line 426, in __init__
    self.prepare_ui(**kwargs)
    ~~~~~~~~~~~~~~~^^^^^^^^^^
  File "/home/rock/Projects/printrun/klimentrun/printrun/stlplater.py", line 210, in prepare_ui
    self.set_viewer(viewer)
    ~~~~~~~~~~~~~~~^^^^^^^^
  File "/home/rock/Projects/printrun/klimentrun/printrun/objectplater.py", line 155, in set_viewer
    if hasattr(viewer.camera, "handle_rotation"):
               ^^^^^^^^^^^^^
AttributeError: 'showstl' object has no attribute 'camera'
```

On the master branch the Plater window is shown and only the expected "no 3D view" error is shown:
```
$ ./pronterface.py 
WARNING:root:Memory-efficient GCoder implementation unavailable: cannot import name 'gcoder_line' from 'printrun' (/home/rock/Projects/printrun/klimentrun/printrun/__init__.py)
3D view mode requested, but we failed to initialize it.
Falling back to 2D view, and here is the backtrace:
Traceback (most recent call last):
  File "/home/rock/Projects/printrun/klimentrun/printrun/gui/viz.py", line 69, in __init__
    import printrun.gcview
  File "/home/rock/Projects/printrun/klimentrun/printrun/gcview.py", line 22, in <module>
    from .gl.panel import wxGLPanel
  File "/home/rock/Projects/printrun/klimentrun/printrun/gl/panel.py", line 29, in <module>
    from pyglet.gl import glEnable, glDisable, GL_LIGHTING, glLightfv, \
    ...<13 lines>...
        GL_LINE_STIPPLE, glColor4f, glLineStipple
ImportError: cannot import name 'GL_LIGHTING' from 'pyglet.gl' (/usr/lib/python3/dist-packages/pyglet/gl/__init__.py)

Could not load 3D viewer for plater:
Traceback (most recent call last):
  File "/home/rock/Projects/printrun/klimentrun/printrun/stlplater.py", line 42, in <module>
    from printrun import stlview
  File "/home/rock/Projects/printrun/klimentrun/printrun/stlview.py", line 25, in <module>
    from pyglet.gl import GL_AMBIENT_AND_DIFFUSE, glBegin, glClearColor, \
    ...<9 lines>...
        GL_LINE_SMOOTH
ImportError: cannot import name 'GL_AMBIENT_AND_DIFFUSE' from 'pyglet.gl' (/usr/lib/python3/dist-packages/pyglet/gl/__init__.py)

Plate function activated
```

DivingDuck commented on printrun/gl/actors.py:
```diff
@@ -1020,11 +1454,14 @@ def load_data(self, model_data, callback=None):
-        logging.debug(_('Initialized 3D visualization in %.2f seconds') % (t_end - t_start))
-        logging.debug(_('Vertex count: %d') % (len(self.vertices) // 3))
+        logging.debug('Initialized 3D visualization in %.2f seconds' % (t_end - t_start))
+        logging.debug('Vertex count: %d' % (len(self.vertices) // 3))
```
Please let this translatable as it already exist

logging.debug(_('Initialized 3D visualization in %.2f seconds') % (t_end - t_start))
        logging.debug(_('Vertex count: %d') % (len(self.vertices) // 3))

DivingDuck commented on printrun/gl/panel.py:
```diff
@@ -110,80 +145,94 @@ def __init__(self, parent, pos = wx.DefaultPosition,
+            except pyglet.gl.GLException:
                 self.gl_broken = True
-                logging.error(_("OpenGL failed, disabling it:")
+                logging.error("OpenGL failed, disabling it:"
```
Please let this translatable as it already exist

 logging.error(_("OpenGL failed, disabling it:")

neofelis2X commented on printrun/gl/actors.py:
```diff
@@ -709,11 +1142,11 @@ def load_data(self, model_data, callback=None):
-        logging.debug(_('Initialized 3D visualization in %.2f seconds') % (t_end - t_start))
-        logging.debug(_('Vertex count: %d') % ((len(self.vertices) + len(self.travels)) // 3))
+        logging.debug('Initialized 3D visualization in %.2f seconds' % (t_end - t_start))
+        logging.debug('Vertex count: %d' % ((len(self.vertices) + len(self.travels)) // 3))
```
Agreed and done.

neofelis2X commented:
Hello @DivingDuck 

> Sorry for my late answer, it was a busy week for me.

No worries, there is no rush! Your additional information helped me to pin down the problems a bit further.

> Mouse scroll wheel for zooming is still not working. Last time I forgot to mention that pan and tilt are working as it is for this version

I was able to reproduce this when I scale down the 3d view to a very small window. I believe _zoom not working_ was caused by a minimum and maximum zoom distance, that I added to the code. I turned this off and hope this will solve this issue.

> For the plater window (stl plater) it still have a problem with handling the stl file in opengl and the fall back in 2D view.

After looking into this I found out that I didn't set the current context before loading a new 3d model. I hope this is fixed now.

> I saw the color of the grid have changed from green to something grayish and recognize that I can't change the color of the grid anymore. It looks like we lost the color setup for this since quite some time (what I thought we had before if I remember correct). 

Yes, the grid had a hard coded colour before. I now changed it in a way that it automatically becomes light grey or dark grey, depending on the colour of the background. Implementing a user picked colour can also be done, if we want that.

> Fit model in window position is not correct (this behavior exists already before)

Yes it is not very good, but it should behave the same way a before. Implementing a better algorithm was not on my todo list for this PR. Too many small details.

So, feel free to test the new binaries, when you find the time. I am optimistic that this two points have been fixed now.

DivingDuck commented:
Hi @neofelis2X,
thanks for the updates. Very nice. Good news, most of my findings are solved. Yay.

> I was able to reproduce this when I scale down the 3d view to a very small window. I believe _zoom not working_ was caused by a minimum and maximum zoom distance, that I added to the code. I turned this off and hope this will solve this issue.

Solved.

> After looking into this I found out that I didn't set the current context before loading a new 3d model. I hope this is fixed now.

Yes. Solved too.

> Yes, the grid had a hard coded colour before. I now changed it in a way that it automatically becomes light grey or dark grey, depending on the colour of the background. Implementing a user picked colour can also be done, if we want that.

This seems not to work. On my computer I can't see the grid anymore. Maybe the color is there but I can't see it.
![#1498Grid](https://github.com/user-attachments/assets/b8dbaaff-53ae-412c-a8dc-bbf8f072d9e6)
I think we should try using the color from the setting options color menu so that all plate colors looking equal and we have a single point for managing those values. Same for all plate dimensions, we should use there as well the values from the settings.

Saying this, I recognize again that we should rework the whole setting story. Is it only me who is stumbling all the time in finding out where a specific default or fix setting was made in our code? I think we should have a central place for those values. Maybe worth or a future project.

Question, is it possible to draw an outer line around the print bed so that one can see what print bed is actual in use (only if it is not a huge effort to do)? 
Background: I have different print beds. Some are rectangular, some are square and one printer have two independent print  beds where I can combine the beds into one big or into independent print areas where I use both or only one of them. Seeing the actual bed frame visual can help preserve me making stupid mistakes (don't ask how do I know).

> Yes it is not very good, but it should behave the same way a before. Implementing a better algorithm was not on my todo list for this PR. Too many small details.

Don't worry, this is somehow as well much better than in my last test. It looks like the zoom problem had forced this glitch.

DivingDuck commented:
@neofelis2X,
I got an other error:
```
Traceback (most recent call last):
  File "C:\Users\Armin\Source\Repos\DivingDuck\Printrun\v3\Lib\site-packages\wx\core.py", line 3425, in <lambda>
    lambda event: event.callable(*event.args, **event.kw) )
                  ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: wrapped C/C++ object of type TempGauge has been deleted
Traceback (most recent call last):
  File "C:\Users\Armin\Source\Repos\DivingDuck\Printrun\v3\Lib\site-packages\wx\core.py", line 3425, in <lambda>
    lambda event: event.callable(*event.args, **event.kw) )
                  ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: wrapped C/C++ object of type TempGauge has been deleted
Traceback (most recent call last):
  File "C:\Users\Armin\Source\Repos\DivingDuck\Printrun\v3\Lib\site-packages\wx\core.py", line 3425, in <lambda>
    lambda event: event.callable(*event.args, **event.kw) )
                  ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Armin\Source\Repos\DivingDuck\Printrun\printrun\gui\graph.py", line 390, in StartPlotting
    self.Refresh()
    ~~~~~~~~~~~~^^
RuntimeError: wrapped C/C++ object of type Graph has been deleted
Traceback (most recent call last):
  File "C:\Users\Armin\Source\Repos\DivingDuck\Printrun\v3\Lib\site-packages\wx\core.py", line 2338, in Notify
    self.notify()
    ~~~~~~~~~~~^^
  File "C:\Users\Armin\Source\Repos\DivingDuck\Printrun\v3\Lib\site-packages\wx\core.py", line 3550, in Notify
    self.result = self.callable(*self.args, **self.kwargs)
                  ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Armin\Source\Repos\DivingDuck\Printrun\printrun\gl\panel.py", line 182, in Refresh
    return super().Refresh(eraseback)
           ~~~~~~~~~~~~~~~^^^^^^^^^^^
RuntimeError: wrapped C/C++ object of type GcodeViewPanel has been deleted
```
The way I force this: Open Pronterface, go to -->Settings -->Options -->Viewer and then for [3D viewer options] check / uncheck both settings for [Use a lighter 3D visualisation] and [Use perspective view instead of orthographic] and then hit the Ok button.
Sometimes this happen only when you restart Pronterface.

neofelis2X commented:
Hi @DivingDuck, so I made some changes to avoid that error.

> Edit: You fixed the grid issue already.

Great, I am glad that this is fine now.  I messed up the colour calculation, but fixed it shortly afterwards.

I will see if I can add the grid colour to the settings.


> Saying this, I recognize again that we should rework the whole setting story. Is it only me who is stumbling all the time in finding out where a specific default or fix setting was made in our code? I think we should have a central place for those values. Maybe worth or a future project.

> I just check this on our latest release and we also have a problem doing this if both entries change at the same time.

Yes, I absolutely agree. The settings are confusing and I try to avoid messing with them, when I can.

This specific error you mentioned can be triggered in many different ways, especially when multiple settings are changed at the same time. The way Pronterface applies changed settings is brute force, it sometimes calls `reload_ui` multiple times in a row.

To improve this a little bit I changed 2 things:
1. Panels and widgets are now destroyed with `DestroyLater` which gives them times to wait for pending events.
2. `reload_ui` now waits for at least 1 ms, to catch any following calls to `reload_ui`. This is not noticeable to the user, but seems to work well.

As you say, generally the whole settings system should probably be reworked. I never quite understood the flow of actions when changing a setting.

rockstorm101 commented:
Hi @neofelis2X,

I've played around a bit and found no show-stoppers. Great job!

> As you say, generally the whole settings system should probably be reworked. I never quite understood the flow of actions when changing a setting.

I agree, it is something I've looked at before. But refactoring the settings system to use something less "homemade" like maybe ConfigParser, wasn't simple. Particularly regarding macros. I could not achieve a solution that wasn't a breaking change for users and/or required a lot of code and effort to keep backwards compatibility.

DivingDuck commented:
Hi @neofelis2X,
I test your last update today: This is solved now. Yay and again great job!

... and I found two other (sorry): 

- When -->Settings -->Options -->Viewer and then for [3D viewer options] is enabled:
Reset view [R] is working fine but Fit [F] is only shifting to the right and not zooming to something for gcode view window/main visualization and fit the plate for stl- and gcode-plater (looks like I do reset view)
I recognize this behavior now the first time as I usually don't have this option enabled. By checking our last releases I see this behavior as well.

- I run plater.py as stand alone version and recognize that there is a problem in loading stl files. You can import one first file but when you try to open a second file the app will go in an infinite loop.

neofelis2X commented:
Hi @rockstorm101, good to hear and thank you very much. I saw only right now that you added some comments, i will look into them.

> I could not achieve a solution that wasn't a breaking change for users and/or required a lot of code and effort to keep backwards compatibility.

Ye I have no doubt that this would be difficult to achieve. Luckily there is no urgent need to change anything.

Hello @DivingDuck, thanks I see we are making progress and even some older bugs get fixed. I think it's good to improve as much as possible, when it can be included easily in this PR. :)

Regarding the [Fit to Frame] shortcut... For the platers it is not implemented. The Stl models do not even have bounding boxes yet, that's why platers just call [Reset view].
The problem with in the main view has probably todo with too big bounding boxes. i will test an example.

Hmm that is strange, seems to happen only on windows. In my pyglet2 branch it does not have any issues with loading. That could be a hint. Otherwise i'm at a loss here for now.

neofelis2X commented:
> I somehow lost to mention: [Use perspective view instead of orthographic] need to be activated for forcing the error

Sorry, which Error? The plater never use perspective view, because this settings is not given to them. Do you mean the [Fit] Error?

neofelis2X commented on printrun/gl/actors.py:
```diff
@@ -105,93 +140,448 @@ def __init__(self, build_dimensions, light = False, circular = False, grid = (1,
-        # glCallList(self.display_list)
-        self.draw()
 
+'''
```
Thinking about it... I will just leave a TODO and delete the rest.

neofelis2X commented on printrun/gl/panel.py:
```diff
@@ -13,92 +13,130 @@
 # BASE_CLASS = wx.Window
 # Subclassing Panel solves problem In Windows
-BASE_CLASS = wx.Panel
+# BASE_CLASS = wx.Panel
```
I removed the comments in the pyglet2 branch. I leave them in here for now, in case anything comes up. Can't really test linux systems that much.

DivingDuck commented:
Hi @neofelis2X,
Indeed, with modernizing/rework the code you fix some of the old errors on the fly :)

For the Fit / Reset view problem:
Here is how I force the error on my system. Open a gcode file, hit Fit[F] and Reset[R]. Then change the setting of  -->Settings -->Options -->Viewer and then toggle for [3D viewer options] the setting of [Use perspective view instead of orthographic] and save the change with [OK]. Hit Fit[F] and Reset[R] again. Now you can see the different behavior between perspective and orthographic view.

![PrintrunOrtjogaphic1](https://github.com/user-attachments/assets/3ebfc637-e5b4-450a-a698-ba93e14df550)

Shame on me regarding the stl and gcode plater issue, I looks like got confused during my tests. I can't reproduce it. You are right

A little fun thing, this week I was searching in the math lib a function and remember a change you did in actors.py regarding 2*pi when I saw math.tau. You can use the constant math.tau instead of 2*math.pi and save the need for a multiplication in line 361, 735, 737 and 956. Maybe worth a little runtime code optimization in remember of the tau day 2025 in June 😇

Regarding the plater.py as stand alone app, it looks like something went wrong and crash the app when it try to open the add model dialog a second time. Interestingly when I run it in visual studio in debug mode it sometimes is successful and sometimes not. In running from code w/o debugging it all times crash when I try to load a second stl file. It looks like there is a timing problem. I also try to run plater as binary file and it sometimes run and more often it hangs after I click the button Add Model for a additional stl file.

DivingDuck commented:
Hi @rockstorm101 , @neofelis2X ,

Sorry that I had no time the last months. 

I think we should release this PR. I haven't a solution for the standalone jet but I guess that most of the users didn't use the standalone version of Plater - at least for Windows and macOS we do not publish a executable binary. We can open an issue to keep that problem in mind for future improvements.
What do you think?
Best regards, DD

PS: Thanks for implementing TAU :)
